core/src/main/java/org/elasticsearch/cluster/service/ClusterService.java

```diff
@@ -571,6 +571,8 @@ public String source() {
                 if (previousClusterState.metaData() != newClusterState.metaData()) {
                     builder.metaData(MetaData.builder(newClusterState.metaData()).version(newClusterState.metaData().version() + 1));
                 }
+                final int numPurged = builder.metaData().indexGraveyard().purge(); // purge old index tombstone entries
```

Comment:
let's keep this out of here - it's complex enough :) see my suggestion to do it on index creation and deletion.